Stub files can no longer undo the class `@NonNull` annotation on, e.g., Optional

Inside Google, we (maybe intentionally, maybe not) have a stub for `java.util.Optional` that undoes [the class `@NonNull` annotation on that type](https://github.com/eisop/jdk/blob/36e42e4279c6d8a47496db8e18983ebbcffcfffe/src/java.base/share/classes/java/util/Optional.java#L86), using [the approach from its `CFComment`](https://github.com/eisop/jdk/blob/36e42e4279c6d8a47496db8e18983ebbcffcfffe/src/java.base/share/classes/java/util/Optional.java#L77-L78). As part of the upgrade from 3.21.4-eisop1 to 3.24.0-eisop1, that stops working. I notice that typetools 3.24.0 still works, so I'm wondering if this is fallout from the eisop `extends @Nullable Object` fixes, similar to https://github.com/eisop/checker-framework/issues/204, or from the eisop stub-file fix of https://github.com/eisop/checker-framework/pull/247. I haven't done any binary searching across releases, though, let alone commits. \[edit: After writing that, I started with `git bisect`. But I ran into trouble because the build process doesn't automatically know which commit of `stubparser` to check out when building older versions.\]

```
$ tail -n +1 X.java mystubs.astub 
==> X.java <==
import org.checkerframework.checker.nullness.qual.Nullable;

interface X {
  java.util.@Nullable Optional<String> o();
}

==> mystubs.astub <==
package java.util;
class Optional<T> {}
```

```
checker-framework-3.21.4-eisop1 $ checker/bin/javac -Astubs=mystubs.astub -processor nullness X.java
warning: /usr/local/google/home/cpovirk/checker-framework-3.21.4-eisop1/mystubs.astub: No supported annotations found! Does stub file /usr/local/google/home/cpovirk/checker-framework-3.21.4-eisop1/mystubs.astub import them?
1 warning
```

```
checker-framework-3.24.0-eisop1 $ checker/bin/javac -Astubs=mystubs.astub -processor nullness X.java
warning: /usr/local/google/home/cpovirk/checker-framework-3.24.0-eisop1/mystubs.astub: No supported annotations found! Does stub file /usr/local/google/home/cpovirk/checker-framework-3.24.0-eisop1/mystubs.astub import them?
X.java:4: error: [type.invalid.annotations.on.use] invalid type: annotations [@Initialized, @Nullable] conflict with declaration of type java.util.Optional
  java.util.@Nullable Optional<String> o();
                                       ^
1 error
1 warning
```

```
checker-framework-3.24.0 $ checker/bin/javac -Astubs=mystubs.astub -processor nullness X.java
warning: /usr/local/google/home/cpovirk/checker-framework-3.24.0/mystubs.astub: No supported annotations found! Does stub file /usr/local/google/home/cpovirk/checker-framework-3.24.0/mystubs.astub import them?
1 warning
```

@amalloy

--- a/checker/jtreg/stubs/stub-over-jdk/Issue321.astub
+++ b/checker/jtreg/stubs/stub-over-jdk/Issue321.astub
@@ -1,0 +1,2 @@
+package java.util;
+class Optional<T> {}

--- a/checker/jtreg/stubs/stub-over-jdk/Issue321.java
+++ b/checker/jtreg/stubs/stub-over-jdk/Issue321.java
@@ -1,0 +1,14 @@
+/*
+ * @test
+ * @summary Test for EISOP Issue #321
+ * @compile/fail/ref=Issue321.out -XDrawDiagnostics -processor org.checkerframework.checker.nullness.NullnessChecker Issue321.java -Anomsgtext
+ * @compile -XDrawDiagnostics -processor org.checkerframework.checker.nullness.NullnessChecker -Astubs=Issue321.astub Issue321.java
+ */
+
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+import java.util.Optional;
+
+interface Issue321 {
+    @Nullable Optional<String> o();
+}

--- a/checker/jtreg/stubs/stub-over-jdk/Issue321.out
+++ b/checker/jtreg/stubs/stub-over-jdk/Issue321.out
@@ -1,1 +1,2 @@
-Issue321.java:12:32: compiler.err.proc.messager: (type.invalid.annotations.on.use)
+Issue321.java:13:32: compiler.err.proc.messager: (type.invalid.annotations.on.use)
+1 error

--- a/checker/jtreg/stubs/stub-over-jdk/Issue321.out
+++ b/checker/jtreg/stubs/stub-over-jdk/Issue321.out
@@ -1,0 +1,1 @@
+Issue321.java:12:32: compiler.err.proc.messager: (type.invalid.annotations.on.use)

--- a/framework/src/main/java/org/checkerframework/framework/stub/AnnotationFileElementTypes.java
+++ b/framework/src/main/java/org/checkerframework/framework/stub/AnnotationFileElementTypes.java
@@ -886,7 +886,7 @@ public class AnnotationFileElementTypes {
         boolean added = processingClasses.add(typeName);
         if (!added) {
             throw new BugInCF(
-                    "Trying to process " + typeName + " which is already being processed.");
+                    "Trying to process type " + typeName + " which is already being processed.");
         }
     }
 
@@ -899,8 +899,7 @@ public class AnnotationFileElementTypes {
     void postProcessTopLevelType(String typeName) {
         boolean removed = processingClasses.remove(typeName);
         if (!removed) {
-            throw new BugInCF(
-                    "Trying to remove a type " + typeName + " that has no processing record.");
+            throw new BugInCF("Cannot find the processing record for type " + typeName);
         }
     }
 }

--- a/framework/src/main/java/org/checkerframework/framework/stub/AnnotationFileElementTypes.java
+++ b/framework/src/main/java/org/checkerframework/framework/stub/AnnotationFileElementTypes.java
@@ -99,21 +99,6 @@ public class AnnotationFileElementTypes {
      */
     private final Set<String> processingClasses = new LinkedHashSet<>();
 
-    private final AnnotationFileParser.Listener afpListener =
-            new AnnotationFileParser.Listener() {
-                @Override
-                public void preProcessTopLevelType(String typeName) {
-                    boolean success = processingClasses.add(typeName);
-                    assert success;
-                }
-
-                @Override
-                public void postProcessTopLevelType(String typeName) {
-                    boolean success = processingClasses.remove(typeName);
-                    assert success;
-                }
-            };
-
     /**
      * Creates an empty annotation source.
      *
@@ -178,7 +163,7 @@ public class AnnotationFileElementTypes {
                         processingEnv,
                         annotationFileAnnos,
                         AnnotationFileType.BUILTIN_STUB,
-                        afpListener);
+                        this);
             }
             String jdkVersionStub = "jdk" + annotatedJdkVersion + ".astub";
             InputStream jdkVersionStubIn = checker.getClass().getResourceAsStream(jdkVersionStub);
@@ -190,7 +175,7 @@ public class AnnotationFileElementTypes {
                         processingEnv,
                         annotationFileAnnos,
                         AnnotationFileType.BUILTIN_STUB,
-                        afpListener);
+                        this);
             }
 
             // 2. Annotated JDK
@@ -255,7 +240,7 @@ public class AnnotationFileElementTypes {
         try {
             InputStream in = new FileInputStream(ajavaPath);
             AnnotationFileParser.parseAjavaFile(
-                    ajavaPath, in, root, factory, processingEnv, annotationFileAnnos, afpListener);
+                    ajavaPath, in, root, factory, processingEnv, annotationFileAnnos, this);
         } catch (IOException e) {
             checker.message(Kind.NOTE, "Could not read ajava file: " + ajavaPath);
         }
@@ -305,7 +290,7 @@ public class AnnotationFileElementTypes {
                             fileType == AnnotationFileType.AJAVA
                                     ? AnnotationFileType.AJAVA_AS_STUB
                                     : fileType,
-                            afpListener);
+                            this);
                 }
             } else {
                 // We didn't find the files.
@@ -317,13 +302,7 @@ public class AnnotationFileElementTypes {
                 InputStream in = checker.getClass().getResourceAsStream(path);
                 if (in != null) {
                     AnnotationFileParser.parseStubFile(
-                            path,
-                            in,
-                            factory,
-                            processingEnv,
-                            annotationFileAnnos,
-                            fileType,
-                            afpListener);
+                            path, in, factory, processingEnv, annotationFileAnnos, fileType, this);
                 } else {
                     // Didn't find the file.  Issue a warning.
 
@@ -722,7 +701,7 @@ public class AnnotationFileElementTypes {
                     factory,
                     factory.getProcessingEnv(),
                     annotationFileAnnos,
-                    afpListener);
+                    this);
         } catch (IOException e) {
             throw new BugInCF("cannot open the jdk stub file " + path, e);
         } finally {
@@ -751,7 +730,7 @@ public class AnnotationFileElementTypes {
                     factory,
                     factory.getProcessingEnv(),
                     annotationFileAnnos,
-                    afpListener);
+                    this);
         } catch (IOException e) {
             throw new BugInCF("cannot open the Jar file " + connection.getEntryName(), e);
         } catch (BugInCF e) {
@@ -895,5 +874,27 @@ public class AnnotationFileElementTypes {
         } catch (IOException e) {
             throw new BugInCF("Cannot open the jar file " + resourceURL.getFile(), e);
         }
+    }
+
+    /**
+     * This method is invoked each time before {@link AnnotationFileParser} processes a top-level
+     * type.
+     *
+     * @param typeName The fully qualified name of the top-level type
+     */
+    void preProcessTopLevelType(String typeName) {
+        boolean success = processingClasses.add(typeName);
+        assert success;
+    }
+
+    /**
+     * This method is invoked each time after {@link AnnotationFileParser} processes a top-level
+     * type.
+     *
+     * @param typeName The fully qualified name of the top-level type
+     */
+    void postProcessTopLevelType(String typeName) {
+        boolean success = processingClasses.remove(typeName);
+        assert success;
     }
 }

--- a/framework/src/main/java/org/checkerframework/framework/stub/AnnotationFileElementTypes.java
+++ b/framework/src/main/java/org/checkerframework/framework/stub/AnnotationFileElementTypes.java
@@ -93,10 +93,26 @@ public class AnnotationFileElementTypes {
     private final boolean parseAllJdkFiles;
 
     /**
-     * Stores the fully qualified name of classes that are currently being parsed in {@link
-     * #parseEnclosingClass}
+     * Stores the fully qualified name of top-level classes (from any type of stub file) that are
+     * currently being parsed. This can stop recursively parsing an annotated JDK class that is
+     * currently being processed, which prevents conflicts of definition and infinite loops.
      */
     private final Set<String> processingClasses = new LinkedHashSet<>();
+
+    private final AnnotationFileParser.Listener afpListener =
+            new AnnotationFileParser.Listener() {
+                @Override
+                public void preProcessTopLevelType(String typeName) {
+                    boolean success = processingClasses.add(typeName);
+                    assert success;
+                }
+
+                @Override
+                public void postProcessTopLevelType(String typeName) {
+                    boolean success = processingClasses.remove(typeName);
+                    assert success;
+                }
+            };
 
     /**
      * Creates an empty annotation source.
@@ -161,7 +177,8 @@ public class AnnotationFileElementTypes {
                         factory,
                         processingEnv,
                         annotationFileAnnos,
-                        AnnotationFileType.BUILTIN_STUB);
+                        AnnotationFileType.BUILTIN_STUB,
+                        afpListener);
             }
             String jdkVersionStub = "jdk" + annotatedJdkVersion + ".astub";
             InputStream jdkVersionStubIn = checker.getClass().getResourceAsStream(jdkVersionStub);
@@ -172,7 +189,8 @@ public class AnnotationFileElementTypes {
                         factory,
                         processingEnv,
                         annotationFileAnnos,
-                        AnnotationFileType.BUILTIN_STUB);
+                        AnnotationFileType.BUILTIN_STUB,
+                        afpListener);
             }
 
             // 2. Annotated JDK
@@ -237,7 +255,7 @@ public class AnnotationFileElementTypes {
         try {
             InputStream in = new FileInputStream(ajavaPath);
             AnnotationFileParser.parseAjavaFile(
-                    ajavaPath, in, root, factory, processingEnv, annotationFileAnnos);
+                    ajavaPath, in, root, factory, processingEnv, annotationFileAnnos, afpListener);
         } catch (IOException e) {
             checker.message(Kind.NOTE, "Could not read ajava file: " + ajavaPath);
         }
@@ -286,7 +304,8 @@ public class AnnotationFileElementTypes {
                             annotationFileAnnos,
                             fileType == AnnotationFileType.AJAVA
                                     ? AnnotationFileType.AJAVA_AS_STUB
-                                    : fileType);
+                                    : fileType,
+                            afpListener);
                 }
             } else {
                 // We didn't find the files.
@@ -298,7 +317,13 @@ public class AnnotationFileElementTypes {
                 InputStream in = checker.getClass().getResourceAsStream(path);
                 if (in != null) {
                     AnnotationFileParser.parseStubFile(
-                            path, in, factory, processingEnv, annotationFileAnnos, fileType);
+                            path,
+                            in,
+                            factory,
+                            processingEnv,
+                            annotationFileAnnos,
+                            fileType,
+                            afpListener);
                 } else {
                     // Didn't find the file.  Issue a warning.
 
@@ -638,20 +663,16 @@ public class AnnotationFileElementTypes {
             return;
         }
 
-        if (!processingClasses.add(className)) {
+        if (processingClasses.contains(className)) {
             // TODO: some declaration annotations in the enclosing class may still
             //  be missing, we can revisit this part if it's causing issues
             return;
         }
 
-        try {
-            if (jdkStubFiles.containsKey(className)) {
-                parseJdkStubFile(jdkStubFiles.remove(className));
-            } else if (jdkStubFilesJar.containsKey(className)) {
-                parseJdkJarEntry(jdkStubFilesJar.remove(className));
-            }
-        } finally {
-            processingClasses.remove(className);
+        if (jdkStubFiles.containsKey(className)) {
+            parseJdkStubFile(jdkStubFiles.remove(className));
+        } else if (jdkStubFilesJar.containsKey(className)) {
+            parseJdkJarEntry(jdkStubFilesJar.remove(className));
         }
     }
 
@@ -700,7 +721,8 @@ public class AnnotationFileElementTypes {
                     jdkStub,
                     factory,
                     factory.getProcessingEnv(),
-                    annotationFileAnnos);
+                    annotationFileAnnos,
+                    afpListener);
         } catch (IOException e) {
             throw new BugInCF("cannot open the jdk stub file " + path, e);
         } finally {
@@ -728,7 +750,8 @@ public class AnnotationFileElementTypes {
                     jdkStub,
                     factory,
                     factory.getProcessingEnv(),
-                    annotationFileAnnos);
+                    annotationFileAnnos,
+                    afpListener);
         } catch (IOException e) {
             throw new BugInCF("cannot open the Jar file " + connection.getEntryName(), e);
         } catch (BugInCF e) {

--- a/framework/src/main/java/org/checkerframework/framework/stub/AnnotationFileElementTypes.java
+++ b/framework/src/main/java/org/checkerframework/framework/stub/AnnotationFileElementTypes.java
@@ -891,7 +891,7 @@ public class AnnotationFileElementTypes {
      * This method is invoked each time after {@link AnnotationFileParser} processes a top-level
      * type.
      *
-     * @param typeName The fully qualified name of the top-level type
+     * @param typeName the fully qualified name of the top-level type
      */
     void postProcessTopLevelType(String typeName) {
         boolean success = processingClasses.remove(typeName);

--- a/framework/src/main/java/org/checkerframework/framework/stub/AnnotationFileElementTypes.java
+++ b/framework/src/main/java/org/checkerframework/framework/stub/AnnotationFileElementTypes.java
@@ -880,7 +880,7 @@ public class AnnotationFileElementTypes {
      * This method is invoked each time before {@link AnnotationFileParser} processes a top-level
      * type.
      *
-     * @param typeName The fully qualified name of the top-level type
+     * @param typeName the fully qualified name of the top-level type
      */
     void preProcessTopLevelType(String typeName) {
         boolean success = processingClasses.add(typeName);

--- a/framework/src/main/java/org/checkerframework/framework/stub/AnnotationFileElementTypes.java
+++ b/framework/src/main/java/org/checkerframework/framework/stub/AnnotationFileElementTypes.java
@@ -883,8 +883,11 @@ public class AnnotationFileElementTypes {
      * @param typeName the fully qualified name of the top-level type
      */
     void preProcessTopLevelType(String typeName) {
-        boolean success = processingClasses.add(typeName);
-        assert success;
+        boolean added = processingClasses.add(typeName);
+        if (!added) {
+            throw new BugInCF(
+                    "Trying to process " + typeName + " which is already being processed.");
+        }
     }
 
     /**
@@ -894,7 +897,10 @@ public class AnnotationFileElementTypes {
      * @param typeName the fully qualified name of the top-level type
      */
     void postProcessTopLevelType(String typeName) {
-        boolean success = processingClasses.remove(typeName);
-        assert success;
+        boolean removed = processingClasses.remove(typeName);
+        if (!removed) {
+            throw new BugInCF(
+                    "Trying to remove a type " + typeName + " that has no processing record.");
+        }
     }
 }

--- a/framework/src/main/java/org/checkerframework/framework/stub/AnnotationFileParser.java
+++ b/framework/src/main/java/org/checkerframework/framework/stub/AnnotationFileParser.java
@@ -126,15 +126,16 @@ import javax.tools.Diagnostic;
  *
  * <p>The first main entry point is {@link #parseStubFile(String, InputStream, AnnotatedTypeFactory,
  * ProcessingEnvironment, AnnotationFileAnnotations, AnnotationFileUtil.AnnotationFileType,
- * Listener)}, which side-effects its last argument. It operates in two steps. First, it calls the
- * Annotation File Parser to parse an annotation file. Then, it walks the AST to create/collect
- * types and declaration annotations.
+ * AnnotationFileElementTypes)}, which side-effects its last argument. It operates in two steps.
+ * First, it calls the Annotation File Parser to parse an annotation file. Then, it walks the AST to
+ * create/collect types and declaration annotations.
  *
  * <p>The second main entry point is {@link #parseAjavaFile(String, InputStream,
  * CompilationUnitTree, AnnotatedTypeFactory, ProcessingEnvironment, AnnotationFileAnnotations,
- * Listener)}. This behaves the same as {@link AnnotationFileParser#parseStubFile(String,
- * InputStream, AnnotatedTypeFactory, ProcessingEnvironment, AnnotationFileAnnotations,
- * AnnotationFileUtil.AnnotationFileType, Listener)}, but takes an ajava file instead.
+ * AnnotationFileElementTypes)}. This behaves the same as {@link
+ * AnnotationFileParser#parseStubFile(String, InputStream, AnnotatedTypeFactory,
+ * ProcessingEnvironment, AnnotationFileAnnotations, AnnotationFileUtil.AnnotationFileType,
+ * AnnotationFileElementTypes)}, but takes an ajava file instead.
  *
  * <p>The other entry point is {@link #parseJdkFileAsStub}.
  */
@@ -205,6 +206,9 @@ public class AnnotationFileParser {
     private final AnnotatedTypeFactory atypeFactory;
     private final Elements elements;
 
+    /** The manager that controls the stub file parsing process. */
+    private final AnnotationFileElementTypes fileElementTypes;
+
     /**
      * The set of annotations found in the file. Keys are both fully-qualified and simple names.
      * There are two entries for each annotation: the annotation's simple name and its
@@ -263,8 +267,6 @@ public class AnnotationFileParser {
 
     /** Whether or not the {@code -AmergeStubsWithSource} command-line argument was passed. */
     private final boolean mergeStubsWithSource;
-
-    private final @Nullable Listener listener;
 
     /**
      * The result of calling AnnotationFileParser.parse: the annotated types and declaration
@@ -410,20 +412,21 @@ public class AnnotationFileParser {
      * @param atypeFactory AnnotatedTypeFactory to use
      * @param processingEnv ProcessingEnvironment to use
      * @param fileType the type of file being parsed (stub file or ajava file) and its source
+     * @param fileElementTypes The manager that controls the stub file parsing process
      */
     private AnnotationFileParser(
             String filename,
             AnnotatedTypeFactory atypeFactory,
             ProcessingEnvironment processingEnv,
             AnnotationFileType fileType,
-            @Nullable Listener listener) {
+            AnnotationFileElementTypes fileElementTypes) {
         this.filename = filename;
         this.atypeFactory = atypeFactory;
         this.processingEnv = processingEnv;
         this.elements = processingEnv.getElementUtils();
         this.fileType = fileType;
         this.root = null;
-        this.listener = listener;
+        this.fileElementTypes = fileElementTypes;
 
         // TODO: This should use SourceChecker.getOptions() to allow
         // setting these flags per checker.
@@ -665,6 +668,7 @@ public class AnnotationFileParser {
      * @param processingEnv ProcessingEnvironment to use
      * @param annotationFileAnnos annotations from the annotation file; side-effected by this method
      * @param fileType the annotation file type and source
+     * @param fileElementTypes The manager that controls the stub file parsing process
      */
     public static void parseStubFile(
             String filename,
@@ -673,9 +677,10 @@ public class AnnotationFileParser {
             ProcessingEnvironment processingEnv,
             AnnotationFileAnnotations annotationFileAnnos,
             AnnotationFileType fileType,
-            @Nullable Listener listener) {
+            AnnotationFileElementTypes fileElementTypes) {
         AnnotationFileParser afp =
-                new AnnotationFileParser(filename, atypeFactory, processingEnv, fileType, listener);
+                new AnnotationFileParser(
+                        filename, atypeFactory, processingEnv, fileType, fileElementTypes);
         try {
             afp.parseStubUnit(inputStream);
             afp.process(annotationFileAnnos);
@@ -696,6 +701,7 @@ public class AnnotationFileParser {
      * @param atypeFactory AnnotatedTypeFactory to use
      * @param processingEnv ProcessingEnvironment to use
      * @param ajavaAnnos annotations from the ajava file; side-effected by this method
+     * @param fileElementTypes The manager that controls the stub file parsing process
      */
     public static void parseAjavaFile(
             String filename,
@@ -704,10 +710,14 @@ public class AnnotationFileParser {
             AnnotatedTypeFactory atypeFactory,
             ProcessingEnvironment processingEnv,
             AnnotationFileAnnotations ajavaAnnos,
-            @Nullable Listener listener) {
+            AnnotationFileElementTypes fileElementTypes) {
         AnnotationFileParser afp =
                 new AnnotationFileParser(
-                        filename, atypeFactory, processingEnv, AnnotationFileType.AJAVA, listener);
+                        filename,
+                        atypeFactory,
+                        processingEnv,
+                        AnnotationFileType.AJAVA,
+                        fileElementTypes);
         try {
             afp.parseStubUnit(inputStream);
             JavaParserUtil.concatenateAddedStringLiterals(afp.stubUnit);
@@ -736,7 +746,7 @@ public class AnnotationFileParser {
             AnnotatedTypeFactory atypeFactory,
             ProcessingEnvironment processingEnv,
             AnnotationFileAnnotations stubAnnos,
-            @Nullable Listener listener) {
+            AnnotationFileElementTypes fileElementTypes) {
         parseStubFile(
                 filename,
                 inputStream,
@@ -744,7 +754,7 @@ public class AnnotationFileParser {
                 processingEnv,
                 stubAnnos,
                 AnnotationFileType.JDK_STUB,
-                listener);
+                fileElementTypes);
     }
 
     /**
@@ -823,18 +833,13 @@ public class AnnotationFileParser {
             if (cu.getTypes() != null) {
                 for (TypeDeclaration<?> typeDeclaration : cu.getTypes()) {
                     Optional<String> typeDeclName = typeDeclaration.getFullyQualifiedName();
-                    boolean callListener = listener != null && typeDeclName.isPresent();
 
-                    if (callListener) {
-                        listener.preProcessTopLevelType(typeDeclName.get());
-                    }
+                    typeDeclName.ifPresent(fileElementTypes::preProcessTopLevelType);
                     try {
                         // Not processing an ajava file, so ignore the return value.
                         processTypeDecl(typeDeclaration, null, null);
                     } finally {
-                        if (callListener) {
-                            listener.postProcessTopLevelType(typeDeclName.get());
-                        }
+                        typeDeclName.ifPresent(fileElementTypes::postProcessTopLevelType);
                     }
                 }
             }
@@ -3141,12 +3146,11 @@ public class AnnotationFileParser {
         public Void visitClass(ClassTree javacTree, Node javaParserNode) {
             TypeDeclaration<?> typeDecl = (TypeDeclaration<?>) javaParserNode;
             Optional<String> typeDeclName = typeDecl.getFullyQualifiedName();
-            boolean callListener =
-                    listener != null && typeDeclName.isPresent() && typeDecl.isTopLevelType();
+            boolean callListener = typeDeclName.isPresent() && typeDecl.isTopLevelType();
             List<AnnotatedTypeVariable> typeDeclTypeParameters = null;
 
             if (callListener) {
-                listener.preProcessTopLevelType(typeDeclName.get());
+                fileElementTypes.preProcessTopLevelType(typeDeclName.get());
             }
             try {
                 if (!(typeDecl instanceof AnnotationDeclaration)) {
@@ -3159,7 +3163,7 @@ public class AnnotationFileParser {
                     typeParameters.removeAll(typeDeclTypeParameters);
                 }
                 if (callListener) {
-                    listener.postProcessTopLevelType(typeDeclName.get());
+                    fileElementTypes.postProcessTopLevelType(typeDeclName.get());
                 }
             }
 

--- a/framework/src/main/java/org/checkerframework/framework/stub/AnnotationFileParser.java
+++ b/framework/src/main/java/org/checkerframework/framework/stub/AnnotationFileParser.java
@@ -417,7 +417,7 @@ public class AnnotationFileParser {
      * @param atypeFactory AnnotatedTypeFactory to use
      * @param processingEnv ProcessingEnvironment to use
      * @param fileType the type of file being parsed (stub file or ajava file) and its source
-     * @param fileElementTypes The manager that controls the stub file parsing process
+     * @param fileElementTypes the manager that controls the stub file parsing process
      */
     private AnnotationFileParser(
             String filename,

--- a/framework/src/main/java/org/checkerframework/framework/stub/AnnotationFileParser.java
+++ b/framework/src/main/java/org/checkerframework/framework/stub/AnnotationFileParser.java
@@ -125,16 +125,16 @@ import javax.tools.Diagnostic;
  * to the {@link AnnotationFileAnnotations} passed as an argument.
  *
  * <p>The first main entry point is {@link #parseStubFile(String, InputStream, AnnotatedTypeFactory,
- * ProcessingEnvironment, AnnotationFileAnnotations, AnnotationFileUtil.AnnotationFileType)}, which
- * side-effects its last argument. It operates in two steps. First, it calls the Annotation File
- * Parser to parse an annotation file. Then, it walks the AST to create/collect types and
- * declaration annotations.
+ * ProcessingEnvironment, AnnotationFileAnnotations, AnnotationFileUtil.AnnotationFileType,
+ * Listener)}, which side-effects its last argument. It operates in two steps. First, it calls the
+ * Annotation File Parser to parse an annotation file. Then, it walks the AST to create/collect
+ * types and declaration annotations.
  *
  * <p>The second main entry point is {@link #parseAjavaFile(String, InputStream,
- * CompilationUnitTree, AnnotatedTypeFactory, ProcessingEnvironment, AnnotationFileAnnotations)}.
- * This behaves the same as {@link AnnotationFileParser#parseStubFile(String, InputStream,
- * AnnotatedTypeFactory, ProcessingEnvironment, AnnotationFileAnnotations,
- * AnnotationFileUtil.AnnotationFileType)}, but takes an ajava file instead.
+ * CompilationUnitTree, AnnotatedTypeFactory, ProcessingEnvironment, AnnotationFileAnnotations,
+ * Listener)}. This behaves the same as {@link AnnotationFileParser#parseStubFile(String,
+ * InputStream, AnnotatedTypeFactory, ProcessingEnvironment, AnnotationFileAnnotations,
+ * AnnotationFileUtil.AnnotationFileType, Listener)}, but takes an ajava file instead.
  *
  * <p>The other entry point is {@link #parseJdkFileAsStub}.
  */
@@ -263,6 +263,8 @@ public class AnnotationFileParser {
 
     /** Whether or not the {@code -AmergeStubsWithSource} command-line argument was passed. */
     private final boolean mergeStubsWithSource;
+
+    private final @Nullable Listener listener;
 
     /**
      * The result of calling AnnotationFileParser.parse: the annotated types and declaration
@@ -413,13 +415,15 @@ public class AnnotationFileParser {
             String filename,
             AnnotatedTypeFactory atypeFactory,
             ProcessingEnvironment processingEnv,
-            AnnotationFileType fileType) {
+            AnnotationFileType fileType,
+            @Nullable Listener listener) {
         this.filename = filename;
         this.atypeFactory = atypeFactory;
         this.processingEnv = processingEnv;
         this.elements = processingEnv.getElementUtils();
         this.fileType = fileType;
         this.root = null;
+        this.listener = listener;
 
         // TODO: This should use SourceChecker.getOptions() to allow
         // setting these flags per checker.
@@ -668,9 +672,10 @@ public class AnnotationFileParser {
             AnnotatedTypeFactory atypeFactory,
             ProcessingEnvironment processingEnv,
             AnnotationFileAnnotations annotationFileAnnos,
-            AnnotationFileType fileType) {
+            AnnotationFileType fileType,
+            @Nullable Listener listener) {
         AnnotationFileParser afp =
-                new AnnotationFileParser(filename, atypeFactory, processingEnv, fileType);
+                new AnnotationFileParser(filename, atypeFactory, processingEnv, fileType, listener);
         try {
             afp.parseStubUnit(inputStream);
             afp.process(annotationFileAnnos);
@@ -698,10 +703,11 @@ public class AnnotationFileParser {
             CompilationUnitTree root,
             AnnotatedTypeFactory atypeFactory,
             ProcessingEnvironment processingEnv,
-            AnnotationFileAnnotations ajavaAnnos) {
+            AnnotationFileAnnotations ajavaAnnos,
+            @Nullable Listener listener) {
         AnnotationFileParser afp =
                 new AnnotationFileParser(
-                        filename, atypeFactory, processingEnv, AnnotationFileType.AJAVA);
+                        filename, atypeFactory, processingEnv, AnnotationFileType.AJAVA, listener);
         try {
             afp.parseStubUnit(inputStream);
             JavaParserUtil.concatenateAddedStringLiterals(afp.stubUnit);
@@ -729,14 +735,16 @@ public class AnnotationFileParser {
             InputStream inputStream,
             AnnotatedTypeFactory atypeFactory,
             ProcessingEnvironment processingEnv,
-            AnnotationFileAnnotations stubAnnos) {
+            AnnotationFileAnnotations stubAnnos,
+            @Nullable Listener listener) {
         parseStubFile(
                 filename,
                 inputStream,
                 atypeFactory,
                 processingEnv,
                 stubAnnos,
-                AnnotationFileType.JDK_STUB);
+                AnnotationFileType.JDK_STUB,
+                listener);
     }
 
     /**
@@ -814,8 +822,20 @@ public class AnnotationFileParser {
         if (fileType.isStub()) {
             if (cu.getTypes() != null) {
                 for (TypeDeclaration<?> typeDeclaration : cu.getTypes()) {
-                    // Not processing an ajava file, so ignore the return value.
-                    processTypeDecl(typeDeclaration, null, null);
+                    Optional<String> typeDeclName = typeDeclaration.getFullyQualifiedName();
+                    boolean callListener = listener != null && typeDeclName.isPresent();
+
+                    if (callListener) {
+                        listener.preProcessTopLevelType(typeDeclName.get());
+                    }
+                    try {
+                        // Not processing an ajava file, so ignore the return value.
+                        processTypeDecl(typeDeclaration, null, null);
+                    } finally {
+                        if (callListener) {
+                            listener.postProcessTopLevelType(typeDeclName.get());
+                        }
+                    }
                 }
             }
         } else {
@@ -3119,16 +3139,28 @@ public class AnnotationFileParser {
     private class AjavaAnnotationCollectorVisitor extends DefaultJointVisitor {
         @Override
         public Void visitClass(ClassTree javacTree, Node javaParserNode) {
+            TypeDeclaration<?> typeDecl = (TypeDeclaration<?>) javaParserNode;
+            Optional<String> typeDeclName = typeDecl.getFullyQualifiedName();
+            boolean callListener =
+                    listener != null && typeDeclName.isPresent() && typeDecl.isTopLevelType();
             List<AnnotatedTypeVariable> typeDeclTypeParameters = null;
-            if (javaParserNode instanceof TypeDeclaration<?>
-                    && !(javaParserNode instanceof AnnotationDeclaration)) {
-                typeDeclTypeParameters =
-                        processTypeDecl((TypeDeclaration<?>) javaParserNode, null, javacTree);
-            }
 
-            super.visitClass(javacTree, javaParserNode);
-            if (typeDeclTypeParameters != null) {
-                typeParameters.removeAll(typeDeclTypeParameters);
+            if (callListener) {
+                listener.preProcessTopLevelType(typeDeclName.get());
+            }
+            try {
+                if (!(typeDecl instanceof AnnotationDeclaration)) {
+                    typeDeclTypeParameters =
+                            processTypeDecl((TypeDeclaration<?>) typeDecl, null, javacTree);
+                }
+                super.visitClass(javacTree, typeDecl);
+            } finally {
+                if (typeDeclTypeParameters != null) {
+                    typeParameters.removeAll(typeDeclTypeParameters);
+                }
+                if (callListener) {
+                    listener.postProcessTopLevelType(typeDeclName.get());
+                }
             }
 
             return null;
@@ -3241,5 +3273,11 @@ public class AnnotationFileParser {
                 return packageName + "." + className;
             }
         }
+    }
+
+    public interface Listener {
+        default void preProcessTopLevelType(String typeName) {}
+
+        default void postProcessTopLevelType(String typeName) {}
     }
 }

--- a/framework/src/main/java/org/checkerframework/framework/stub/AnnotationFileParser.java
+++ b/framework/src/main/java/org/checkerframework/framework/stub/AnnotationFileParser.java
@@ -3278,10 +3278,4 @@ public class AnnotationFileParser {
             }
         }
     }
-
-    public interface Listener {
-        default void preProcessTopLevelType(String typeName) {}
-
-        default void postProcessTopLevelType(String typeName) {}
-    }
 }

--- a/framework/src/main/java/org/checkerframework/framework/stub/AnnotationFileParser.java
+++ b/framework/src/main/java/org/checkerframework/framework/stub/AnnotationFileParser.java
@@ -673,7 +673,7 @@ public class AnnotationFileParser {
      * @param processingEnv ProcessingEnvironment to use
      * @param annotationFileAnnos annotations from the annotation file; side-effected by this method
      * @param fileType the annotation file type and source
-     * @param fileElementTypes The manager that controls the stub file parsing process
+     * @param fileElementTypes the manager that controls the stub file parsing process
      */
     public static void parseStubFile(
             String filename,

--- a/framework/src/main/java/org/checkerframework/framework/stub/AnnotationFileParser.java
+++ b/framework/src/main/java/org/checkerframework/framework/stub/AnnotationFileParser.java
@@ -744,6 +744,7 @@ public class AnnotationFileParser {
      * @param atypeFactory AnnotatedTypeFactory to use
      * @param processingEnv ProcessingEnvironment to use
      * @param stubAnnos annotations from the stub file; side-effected by this method
+     * @param fileElementTypes the manager that controls the stub file parsing process
      */
     public static void parseJdkFileAsStub(
             String filename,
@@ -3149,20 +3150,28 @@ public class AnnotationFileParser {
     private class AjavaAnnotationCollectorVisitor extends DefaultJointVisitor {
         @Override
         public Void visitClass(ClassTree javacTree, Node javaParserNode) {
-            TypeDeclaration<?> typeDecl = (TypeDeclaration<?>) javaParserNode;
-            Optional<String> typeDeclName = typeDecl.getFullyQualifiedName();
-            boolean callListener = typeDeclName.isPresent() && typeDecl.isTopLevelType();
             List<AnnotatedTypeVariable> typeDeclTypeParameters = null;
+            boolean shouldProcessTypeDecl =
+                    javaParserNode instanceof TypeDeclaration<?>
+                            && !(javaParserNode instanceof AnnotationDeclaration);
+            Optional<String> typeDeclName = Optional.empty();
+            boolean callListener = false;
+
+            if (shouldProcessTypeDecl) {
+                TypeDeclaration<?> typeDecl = (TypeDeclaration<?>) javaParserNode;
+                typeDeclName = typeDecl.getFullyQualifiedName();
+                callListener = typeDeclName.isPresent() && typeDecl.isTopLevelType();
+            }
 
             if (callListener) {
                 fileElementTypes.preProcessTopLevelType(typeDeclName.get());
             }
             try {
-                if (!(typeDecl instanceof AnnotationDeclaration)) {
+                if (shouldProcessTypeDecl) {
                     typeDeclTypeParameters =
-                            processTypeDecl((TypeDeclaration<?>) typeDecl, null, javacTree);
+                            processTypeDecl((TypeDeclaration<?>) javaParserNode, null, javacTree);
                 }
-                super.visitClass(javacTree, typeDecl);
+                super.visitClass(javacTree, javaParserNode);
             } finally {
                 if (typeDeclTypeParameters != null) {
                     typeParameters.removeAll(typeDeclTypeParameters);

--- a/framework/src/main/java/org/checkerframework/framework/stub/AnnotationFileParser.java
+++ b/framework/src/main/java/org/checkerframework/framework/stub/AnnotationFileParser.java
@@ -202,8 +202,13 @@ public class AnnotationFileParser {
     // Not final in order to accommodate a default value.
     private StubUnit stubUnit;
 
+    /** The processing environment */
     private final ProcessingEnvironment processingEnv;
+
+    /** The type factory */
     private final AnnotatedTypeFactory atypeFactory;
+
+    /** The element utilities */
     private final Elements elements;
 
     /** The manager that controls the stub file parsing process. */

--- a/framework/src/main/java/org/checkerframework/framework/stub/AnnotationFileParser.java
+++ b/framework/src/main/java/org/checkerframework/framework/stub/AnnotationFileParser.java
@@ -706,7 +706,7 @@ public class AnnotationFileParser {
      * @param atypeFactory AnnotatedTypeFactory to use
      * @param processingEnv ProcessingEnvironment to use
      * @param ajavaAnnos annotations from the ajava file; side-effected by this method
-     * @param fileElementTypes The manager that controls the stub file parsing process
+     * @param fileElementTypes the manager that controls the stub file parsing process
      */
     public static void parseAjavaFile(
             String filename,

